### PR TITLE
new: server-side RSMM support

### PR DESCRIPTION
**Pull request by SpaceWalkerRS**

### CODE REVIEW

**Comment by Fallen-Breath** [comment]
Man that's a lot of codes xd

Before reviewing, I want to know:

- What parts of minecraft codes did you change?
  - Where do your codes hook at? e.g. for blocks's event, or for game phases timing
  - What new methods have you added to which classes? For what?
  - Any new network protocol thing? If there is, is it safe and when / how much will the data be sent
- What parts of carpet codes did you change? rule? carpet protocols? Does it support switching on/off in game?
- Where's is your rsmm impl codes? All in `redstone.multimeter`?
- Does it have performance impact when it's off, what about when it's on?
- Any permission control for players?
- Any file-wise operation? e.g. configure file

tl;dr please give more information of this PR

**Comment by SpaceWalkerRS** [comment]
- RSMM hooks into the MC code in a few places:
  - `World`, `WorldServer`, `MinecraftServer` and `IntegratedServer` for keeping track of tick phases.
  - `World`, `WorldServer`, `Chunk`, `IBlockState` and `IFluidState` to capture block use, block update, comparator update, shape update, random tick, scheduled tick, block event, entity tick and block entity tick events as well as blockstate changes.
  - `Block` and several other block classes, for two purposes:
    - Adding public methods for getting information on whether a block is being powered, active, etc.
    - Adding to/adjusting a method to capture a local variable. This applies to blocks like `BlockPistonBase`, `BlockFenceGate`, `BlockHopper` (among others) where RSMM captures the outcome of the power check so it doesn't have to do an extra power check itself (this saves a bit on performance). 
  - `NetHandlerPlayServer` to handle packets:
    - A "handshake" with any clients that join that have RSMM installed.
    - At the end of each tick, the server sends a packet to all RSMM clients with the current server time.
    - Meter logs - logs of all the events RSMM captures. Each meter group can send this packet, and it is sent at most once per tick, but only if there are any logs.
    - Meter updates - added/removed meters, color changes, name changes, etc. Similar to the meter logs, each meter group can send this packet, at most once per tick, but only if there are any changes.
    - The server also receives packets from RSMM clients to handle meter updates, meter group subscriptions, etc.
  - `Commands` to register the `/metergroup` command.
  - `PlayerList` to capture players joining/leaving.
- I added a carpet rule to `CarpetSettings`. When it is switched off, RSMM stops logging any events, stops listening to packets, and when the command is used it prompts you to enable to carpet rule before you can use it. Performance impact when the rule is enabled is already small, but with it disabled it is negligible. 
- The rest of the RSMM code is in the `redstone.multimeter` package.
- There is a `options.json` file in `/config/redstone_multimeter/` with some (options)[https://github.com/SpaceWalkerRS/redstone-multimeter-fabric/wiki/Server-Options] for dedicated servers.

File: patches/net/minecraft/block/BlockRedstoneLamp.java.patch [hunk 1]
```diff
@@ -0,0 +1,32 @@
...
+-public class BlockRedstoneLamp extends Block
++public class BlockRedstoneLamp extends Block implements MeterableBlock /*RSMM*/
+ {
+     public static final BooleanProperty LIT = BlockRedstoneTorch.LIT;
+ 
+@@ -59,6 +60,7 @@
+     {
+         if (!worldIn.isRemote)
+         {
++            logPowered(worldIn, pos, isPowered(worldIn, pos, state)); // RSMM
```
Comment (Fallen-Breath) [inline patches/net/minecraft/block/BlockRedstoneLamp.java.patch]: `isPowered` is invoked without rule check so potential performance impact

File: patches/net/minecraft/block/BlockPistonBase.java.patch [hunk 2]
```diff
@@ -169,7 +184,15 @@
...
              for (int l = list.size() - 1; l >= 0; --l)
              {
-@@ -379,7 +442,22 @@
+                 BlockPos blockpos3 = list.get(l);
+                 IBlockState iblockstate3 = worldIn.getBlockState(blockpos3);
++                // RSMM start
++                if (!worldIn.isRemote()) {
++                    Multimeter multimeter = ((WorldServer)worldIn).getMultimeter();
++                    multimeter.logMoved(worldIn, blockpos3, enumfacing);
++                    multimeter.moveMeters(worldIn, blockpos3, enumfacing);
```
Comment (Fallen-Breath) [inline patches/net/minecraft/block/BlockPistonBase.java.patch:193]: no rule check here too and there are hashmap scanning in `moveMeters`

**Comment by Fallen-Breath** [comment]
Generally speaking, I want have the part of rsmm codes untouched as much as possible when the rule is off, for performance and for safety (in case exception occurs). 
I have seen u make the rule check before event logging in the `tryLogEvent` thing, but there are still some other rsmm codes that runs without rule check. Simple operations are fine tho.
Maybe you can use the rule check to wrap your codes more I think? Everthing else seems fine to me for now

**Comment by SpaceWalkerRS** [comment]
Yep, you're right. There are a few places where code runs unnecessarily that I missed. I could potentially also add some extra checks "higher up" to limit method calls. I will make some changes when I get back to my Windows machine this weekend.

**Review by Fallen-Breath — APPROVED** [review]

**Comment by Fallen-Breath** [comment]
Everything is fine to me now, just resoloved the conflict

**Comment by Fallen-Breath** [comment]
Ah wait, seems like RSMM provides a `/metergroup` command, related permission control might be necessary (`SettingsManager.canUseCommand` at the root command node) so no one see this command when the rule is off. 

idk if it's necessary to change type of rule `redstoneMultimeter` to `String` like other `commandXxx` since it doesn't make sense to only allow part of players to use it

So maybe this should be ok

```java
literal("metergroup").
requires(source -> CarpetSettings.redstoneMultimeter && isMultimeterClient(source)).
```

I can make the modification myself if you can't get your Windows machine now

**Comment by SpaceWalkerRS** [comment]
Are the clients' command permissions updated when carpet rules are changed? I think I assumed that wasn't the case and didn't include the rule check in the permission check for that reason. 

> I can make the modification myself if you can't get your Windows machine now

Sounds good. I will be away from my Windows machine for a few more days so go ahead.

**Comment by Fallen-Breath** [comment]
I only saw u added this `redstoneMultimeter` rule, but no rule like `commandMeterGroup` so i'm assuming u being use the `redstoneMultimeter` rule for the global switch of rsmm

for the command sync thing, u need to add COMMAND to the rule's catagory to let carpet to handle the commabd tree change due to permission change

**Comment by SpaceWalkerRS** [comment]
Yeah, I did intend for that one rule to turn everything RSMM related on/off. What I did for the command was to have it fail and prompt the user to enable to carpet rule if it weren't enabled, but that is not exactly ideal. 

> u need to add COMMAND to the rule's catagory

I will do that and then remove the (then) useless prompt to enable the rule from the command feedback.
